<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Science for Biology - Introduction to R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/DRP_Logo_only_square_transparent.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Data Science for Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./01-start.html"> 
<span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r-an-rstudio" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">R an RStudio</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r-an-rstudio">    
        <li>
    <a class="dropdown-item" href="./02-Setup.html">
 <span class="dropdown-text">Course setup</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-IntroR.html">
 <span class="dropdown-text">Introduction to R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-Programming.html">
 <span class="dropdown-text">Programming in R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-HW.html">
 <span class="dropdown-text">Homework task</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#what-is-r" id="toc-what-is-r" class="nav-link active" data-scroll-target="#what-is-r">What is R?</a></li>
  <li><a href="#why-is-r-useful" id="toc-why-is-r-useful" class="nav-link" data-scroll-target="#why-is-r-useful">Why is R useful?</a></li>
  <li><a href="#working-directory" id="toc-working-directory" class="nav-link" data-scroll-target="#working-directory">Working directory</a></li>
  <li><a href="#data-types" id="toc-data-types" class="nav-link" data-scroll-target="#data-types">Data Types</a></li>
  <li><a href="#data-structures" id="toc-data-structures" class="nav-link" data-scroll-target="#data-structures">Data Structures</a>
  <ul class="collapse">
  <li><a href="#vector" id="toc-vector" class="nav-link" data-scroll-target="#vector">Vector</a></li>
  <li><a href="#data-frame" id="toc-data-frame" class="nav-link" data-scroll-target="#data-frame">Data Frame</a></li>
  <li><a href="#matrix" id="toc-matrix" class="nav-link" data-scroll-target="#matrix">Matrix</a></li>
  <li><a href="#list" id="toc-list" class="nav-link" data-scroll-target="#list">List</a></li>
  </ul></li>
  <li><a href="#r-syntax" id="toc-r-syntax" class="nav-link" data-scroll-target="#r-syntax">R Syntax</a></li>
  <li><a href="#storing-objects" id="toc-storing-objects" class="nav-link" data-scroll-target="#storing-objects">Storing objects</a></li>
  <li><a href="#getting-help" id="toc-getting-help" class="nav-link" data-scroll-target="#getting-help">Getting help</a></li>
  <li><a href="#getting-data-into-r" id="toc-getting-data-into-r" class="nav-link" data-scroll-target="#getting-data-into-r">Getting data into R</a>
  <ul class="collapse">
  <li><a href="#other-options-for-read.table" id="toc-other-options-for-read.table" class="nav-link" data-scroll-target="#other-options-for-read.table">Other options for read.table</a></li>
  <li><a href="#importing-datasets" id="toc-importing-datasets" class="nav-link" data-scroll-target="#importing-datasets">Importing Datasets</a></li>
  </ul></li>
  <li><a href="#inspect-and-summarise-data" id="toc-inspect-and-summarise-data" class="nav-link" data-scroll-target="#inspect-and-summarise-data">Inspect and summarise data</a></li>
  <li><a href="#square-bracket-notation" id="toc-square-bracket-notation" class="nav-link" data-scroll-target="#square-bracket-notation">Square bracket notation</a></li>
  <li><a href="#ordering-data" id="toc-ordering-data" class="nav-link" data-scroll-target="#ordering-data">Ordering data</a></li>
  <li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data">Filtering data</a></li>
  <li><a href="#plotting-with-r" id="toc-plotting-with-r" class="nav-link" data-scroll-target="#plotting-with-r">Plotting with R</a>
  <ul class="collapse">
  <li><a href="#histograms" id="toc-histograms" class="nav-link" data-scroll-target="#histograms">Histograms</a></li>
  <li><a href="#boxplots" id="toc-boxplots" class="nav-link" data-scroll-target="#boxplots">Boxplots</a></li>
  <li><a href="#scatter-plots" id="toc-scatter-plots" class="nav-link" data-scroll-target="#scatter-plots">Scatter plots</a></li>
  </ul></li>
  <li><a href="#saving-images" id="toc-saving-images" class="nav-link" data-scroll-target="#saving-images">Saving images</a>
  <ul class="collapse">
  <li><a href="#use-the-export-button-in-the-plots-pane-in-rstudio." id="toc-use-the-export-button-in-the-plots-pane-in-rstudio." class="nav-link" data-scroll-target="#use-the-export-button-in-the-plots-pane-in-rstudio.">Use the <strong>export</strong> button in the Plots pane in Rstudio.</a></li>
  <li><a href="#use-a-graphics-device-function-in-your-r-code." id="toc-use-a-graphics-device-function-in-your-r-code." class="nav-link" data-scroll-target="#use-a-graphics-device-function-in-your-r-code.">Use a <strong>graphics device</strong> function in your R code.</a></li>
  </ul></li>
  <li><a href="#statistical-testing" id="toc-statistical-testing" class="nav-link" data-scroll-target="#statistical-testing">Statistical testing</a>
  <ul class="collapse">
  <li><a href="#t-test" id="toc-t-test" class="nav-link" data-scroll-target="#t-test">T-test</a></li>
  <li><a href="#fishers-exact-test" id="toc-fishers-exact-test" class="nav-link" data-scroll-target="#fishers-exact-test">Fisher’s exact test</a></li>
  </ul></li>
  <li><a href="#case-study-quantitative-pcr" id="toc-case-study-quantitative-pcr" class="nav-link" data-scroll-target="#case-study-quantitative-pcr">Case Study: Quantitative PCR</a>
  <ul class="collapse">
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">Interpretation</a></li>
  <li><a href="#reference-gene-normalisation" id="toc-reference-gene-normalisation" class="nav-link" data-scroll-target="#reference-gene-normalisation">Reference gene normalisation</a></li>
  <li><a href="#comparing-conditions" id="toc-comparing-conditions" class="nav-link" data-scroll-target="#comparing-conditions">Comparing conditions</a></li>
  <li><a href="#step-1-load-the-data" id="toc-step-1-load-the-data" class="nav-link" data-scroll-target="#step-1-load-the-data">Step 1: Load the data</a></li>
  <li><a href="#step-2-calculate-δct" id="toc-step-2-calculate-δct" class="nav-link" data-scroll-target="#step-2-calculate-δct">Step 2: Calculate ΔCt</a></li>
  <li><a href="#step-3-visualise-the-data" id="toc-step-3-visualise-the-data" class="nav-link" data-scroll-target="#step-3-visualise-the-data">Step 3: Visualise the data</a></li>
  <li><a href="#step-4-perform-a-t-test" id="toc-step-4-perform-a-t-test" class="nav-link" data-scroll-target="#step-4-perform-a-t-test">Step 4: Perform a t-test</a></li>
  <li><a href="#step-5-calculate-fold-change" id="toc-step-5-calculate-fold-change" class="nav-link" data-scroll-target="#step-5-calculate-fold-change">Step 5: Calculate fold change</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction to R</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>



<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>

<div class="objectives">
<h2 class="anchored">
<i class="far fa-check-square"></i> Learning Objectives
</h2>
<ul>
<li>What is R?</li>
<li>Understand R datatypes and classes</li>
<li>Understand how to use functions in <em>base R</em></li>
<li>Import datasets into R</li>
<li>Inspect and format a dataset</li>
<li>Create basic graphics</li>
<li>Run statistical tests</li>
</ul>
</div>
<p><br></p>
<section id="what-is-r" class="level2">
<h2 class="anchored" data-anchor-id="what-is-r">What is R?</h2>
<p><a href="http://www.r-project.org/">R</a> is an extremely powerful programming language for data science, statistical analysis and data visualisation. It is favoured by biologists due to its extensive library of packages for bioinformatics and genomic analysis. In this lesson you will learn how to program and interact with data in R. We will cover the basics of the R syntax, using its built in packages (base R).</p>
</section>
<section id="why-is-r-useful" class="level2">
<h2 class="anchored" data-anchor-id="why-is-r-useful">Why is R useful?</h2>
<ul>
<li><p>It’s <strong>free!</strong> open source software.</p></li>
<li><p>It’s <strong>Powerful</strong>. There are many libraries for application specific tasks.</p>
<ul>
<li><p><a href="https://www.bioconductor.org/">Bioconductor</a> is a repository for bioinformatics R software.</p></li>
<li><p><a href="https://www.tidyverse.org/">Tidyverse</a> is a suite of packages for data science with a shared philosophy for storing, manipulating and visualising data.</p></li>
</ul></li>
<li><p>Presentation quality graphics</p>
<ul>
<li>Save as a png, pdf or svg for use in publications and presentations.</li>
</ul></li>
<li><p>Generate reproducible and persistent results</p>
<ul>
<li>R commands and analyses can be saved for reproducible and persistent analyses.</li>
<li>Create automated scripts to replicate analysis on different datasets.</li>
</ul></li>
<li><p><a href="https://www.rstudio.com/">RStudio</a> provides an interactive environment for working in R.</p></li>
<li><p><a href="https://rmarkdown.rstudio.com/">R markdown</a> can generate documents to present your analyses as reports.</p></li>
<li><p><a href="https://shiny.rstudio.com/">Shiny</a> can produce interactive applications for exploratory data analysis.</p></li>
</ul>
</section>
<section id="working-directory" class="level2">
<h2 class="anchored" data-anchor-id="working-directory">Working directory</h2>
<p>This is the directory used to store your data and results. This is where R will look for files to read in and where it will save any output files. You can check your current working directory with the <code>getwd()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you want to change the working directory you can use <code>setwd("/path/to/new_directory")</code> or use the <em>session</em> menu in RStudio.</p>
</section>
<section id="data-types" class="level2">
<h2 class="anchored" data-anchor-id="data-types">Data Types</h2>
<p>Values in R are assigned a data type which tells R how to interpret them. Some common data types are:</p>
<ul>
<li><strong>numeric</strong>: Numbers</li>
<li><strong>character</strong>: Strings of text</li>
<li><strong>factor</strong>: Labels for categorical data (e.g.&nbsp;species, sex)</li>
<li><strong>logical</strong>: TRUE or FALSE values</li>
</ul>
</section>
<section id="data-structures" class="level2">
<h2 class="anchored" data-anchor-id="data-structures">Data Structures</h2>
<p>R uses different data structures to organise data. Common structures are shown below.</p>
<section id="vector" class="level3">
<h3 class="anchored" data-anchor-id="vector">Vector</h3>
<ul>
<li>A collection of values of one data type</li>
<li>Equivalent to a column in a table</li>
<li>Items in the vector can be named</li>
<li>E.g. A collection of recorded mouse weights</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## numeric vector</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fl">12.3</span>, <span class="fl">15.6</span>, <span class="fl">9.8</span>, <span class="fl">7.4</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12.3 15.6  9.8  7.4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## A named vector</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="at">M1=</span><span class="fl">12.3</span>, <span class="at">M2=</span><span class="fl">15.6</span>, <span class="at">M3=</span><span class="fl">9.8</span>, <span class="at">M4=</span><span class="fl">7.4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  M1   M2   M3   M4 
12.3 15.6  9.8  7.4 </code></pre>
</div>
</div>
</section>
<section id="data-frame" class="level3">
<h3 class="anchored" data-anchor-id="data-frame">Data Frame</h3>
<ul>
<li>A table</li>
<li>Essentially a collection of vectors as columns</li>
<li>Columns can be different data types</li>
<li>Columns must have the same length</li>
<li>E.g. A table of mouse weights with columns (Mouse_ID (character), Sex (factor), Weight (numeric))</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Mouse_ID =</span> <span class="fu">c</span>(<span class="st">"M1"</span>, <span class="st">"M2"</span>, <span class="st">"M3"</span>),</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Sex =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"M"</span>, <span class="st">"F"</span>, <span class="st">"F"</span>)), <span class="co"># use the factor function to tell R that this is categorical data</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Weight =</span> <span class="fu">c</span>(<span class="fl">12.3</span>, <span class="fl">15.6</span>, <span class="fl">9.8</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Mouse_ID Sex Weight
1       M1   M   12.3
2       M2   F   15.6
3       M3   F    9.8</code></pre>
</div>
</div>
</section>
<section id="matrix" class="level3">
<h3 class="anchored" data-anchor-id="matrix">Matrix</h3>
<ul>
<li>A table where all values are related and of the same data type</li>
<li>Commonly used for correlation and heatmap analysis</li>
<li>E.g. A table of RNA-seq expression levels where each row is a gene and each column is a different sample.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">8</span>, <span class="dv">7</span>, <span class="dv">4</span>), <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>, <span class="st">"Sample3"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2 Sample3
GeneA       5       3       7
GeneB       2       8       4</code></pre>
</div>
</div>
</section>
<section id="list" class="level3">
<h3 class="anchored" data-anchor-id="list">List</h3>
<ul>
<li>Lists are collections of R objects.</li>
<li>Each item in the list has a unique index or name.</li>
<li>A list can contain items of different object types and classes (e.g single values, vectors, data frames, matrices, other lists…).</li>
<li>E.g. A list containing a data frame of sample information, a matrix of expression values and a vector of gene names.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_info =</span> <span class="fu">data.frame</span>(<span class="at">SampleID=</span><span class="fu">c</span>(<span class="st">"Sample1"</span>,<span class="st">"Sample2"</span>), <span class="at">Condition=</span><span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Control"</span>,<span class="st">"Treated"</span>))), <span class="co"># data frame</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression_matrix =</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">8</span>), <span class="at">nrow=</span><span class="dv">2</span>), <span class="co"># matrix of expression values</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_names =</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>) <span class="co"># vector</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$sample_info
  SampleID Condition
1  Sample1   Control
2  Sample2   Treated

$expression_matrix
     [,1] [,2]
[1,]    5    3
[2,]    2    8

$gene_names
[1] "GeneA" "GeneB"</code></pre>
</div>
</div>
</section>
</section>
<section id="r-syntax" class="level2">
<h2 class="anchored" data-anchor-id="r-syntax">R Syntax</h2>
<p>R is a <strong>functional</strong> programming language:</p>
<ul>
<li>Nearly every command is the name of a function followed by parentheses.</li>
<li>The inputs to a function, including different options, are placed in the brackets.</li>
<li>You can use the <kbd>Tab</kbd> key to see the options available or use the help documentation for each function.</li>
</ul>
<p>Typical command structure:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">function_name</span>(data, options, moreOptions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s run a function on a real data set. The <code>cars</code> data set is built into base R. We can look at it by typing its name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>cars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   speed dist
1      4    2
2      4   10
3      7    4
4      7   22
5      8   16
6      9   10
7     10   18
8     10   26
9     10   34
10    11   17
11    11   28
12    12   14
13    12   20
14    12   24
15    12   28
16    13   26
17    13   34
18    13   34
19    13   46
20    14   26
21    14   36
22    14   60
23    14   80
24    15   20
25    15   26
26    15   54
27    16   32
28    16   40
29    17   32
30    17   40
31    17   50
32    18   42
33    18   56
34    18   76
35    18   84
36    19   36
37    19   46
38    19   68
39    20   32
40    20   48
41    20   52
42    20   56
43    20   64
44    22   66
45    23   54
46    24   70
47    24   92
48    24   93
49    24  120
50    25   85</code></pre>
</div>
</div>
<p>We can see that this is a table of values. If we run the <code>class</code> function we will see that R recognises this as a <code>data.frame</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(cars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<p>Now let’s run a more useful function. The <code>plot</code> function in R is used for making basic graphs. It also has additional arguments which can be altered to change its behaviour. Try plotting the cars data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Use the options within the plot function to customise the output</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cars, <span class="at">xlab =</span> <span class="st">"Car Speed (mph)"</span>, <span class="at">ylab =</span> <span class="st">"Stopping Distance (ft)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Not all functions need arguments. For instance, the <code>getwd</code> function which returns the path of your working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="storing-objects" class="level2">
<h2 class="anchored" data-anchor-id="storing-objects">Storing objects</h2>
<p>We can create objects in R to store data or the output of functions. This is useful for saving data that we want to use later on, or for storing intermediate steps in an analysis.</p>
<p>We can use the <code>=</code> sign or <code>&lt;-</code> to store the output of a function as an object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## These statements are identical</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">function_name</span>(data, options, moreOptions)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">function_name</span>(data, options, moreOptions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Try storing the output of the <code>summary</code> function on the <code>cars</code> data set. You will see the object <code>sum_cars</code> appear in your Environment tab in RStudio.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sum_cars <span class="ot">=</span> <span class="fu">summary</span>(cars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To see what this object holds, just type its name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sum_cars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     speed           dist       
 Min.   : 4.0   Min.   :  2.00  
 1st Qu.:12.0   1st Qu.: 26.00  
 Median :15.0   Median : 36.00  
 Mean   :15.4   Mean   : 42.98  
 3rd Qu.:19.0   3rd Qu.: 56.00  
 Max.   :25.0   Max.   :120.00  </code></pre>
</div>
</div>
</section>
<section id="getting-help" class="level2">
<h2 class="anchored" data-anchor-id="getting-help">Getting help</h2>
<p>R has extensive built in help documentation. You can access this using the <code>help</code> function. The command below provides the help page for the <em>function</em> <code>read.table</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(read.table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also search the help documentation using <code>help.search</code>. Let’s see if we can find a function for running a t-test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help.search</span>(<span class="st">"t test"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>NOTE</strong> quotes are required around strings of text. You do not need to use quotes when referring to names of R data objects or functions.</p>
<p>There is a short cut for help, <code>?</code>, which shows the help page for a function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same as help(read.table)</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>?read.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>??</code> searches for help pages on functions, same as help.search(“phrase”)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same as help.search("t test")</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>??<span class="st">"t test"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can also use the <strong>Help</strong> tab in RStudio to search for help on functions, or search online.</p>
<div class="key-points">
<h2 class="anchored">
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li><strong>Data types</strong>: Understand different data types (numeric, character, factor, logical)</li>
<li><strong>R objects</strong>: Understand different ways to structure data in R (vectors, dataframes, matrices, lists). There are several other types of R object.</li>
<li><strong>Functions</strong>: Code in R is run within functions.</li>
<li><strong>Help</strong>: Use the help features to find out how a function works.</li>
</ul>
</div>
</section>
<section id="getting-data-into-r" class="level2">
<h2 class="anchored" data-anchor-id="getting-data-into-r">Getting data into R</h2>
<p>Getting data into R is often the first step in any analysis.</p>
<p>It is possible to create a vector of data by typing directly into R using the function <code>c</code>. The <em>c</em> stands for <em>combine</em> and can be used to combine values into a <em>vector</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>x   <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This creates a vector named ‘x’ which stores the numbers 1 through to 5.</p>
<p>You can see what is in an object at any time by typing its name:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 2 3 4 5</code></pre>
</div>
</div>
<p>Character values need to be quoted, otherwise R will look for a data object of that name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This produces an error because R is looking for objects called Monday, Tuesday etc. which do not exist.</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>daysofweek <span class="ot">&lt;-</span> <span class="fu">c</span>(Monday, Tuesday, Wednesday, Thursday, Friday)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This works</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>daysofweek <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Monday"</span>, <span class="st">"Tuesday"</span>, <span class="st">"Wednesday"</span>, <span class="st">"Thursday"</span>, <span class="st">"Friday"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can create data frames from vectors using the <code>data.frame</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Index =</span> x, <span class="at">Day =</span> daysofweek)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Index       Day
1     1    Monday
2     2   Tuesday
3     3 Wednesday
4     4  Thursday
5     5    Friday</code></pre>
</div>
</div>
<p>Usually however, you will want to input data from a file. You can read files on your computer or from a URL. Use the <code>read.table</code> function to read in the table hosted at the URL below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read.table</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/data.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R reads this table as a <code>data.frame</code> object and prints it to the R console by default. To save the table, we need to assign it to an object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>mydata  <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/data.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, <code>mydata</code> is an object name and the syntax <code>&lt;-</code> assigns the output of the function to it. Remember you can also use <code>=</code>.</p>
<p>R stores <code>mydata</code> as a data frame, containing multiple vectors. We can check the class of our object using the <code>class()</code> function. The <em>Environment</em> tab in RStudio can also be used to explore objects and their properties.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<p>We can look at our table by typing its name, but this prints a lot of rows. Using the <code>head()</code> function will only print the first few lines;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mydata, <span class="at">n=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  V1 V2 V3 V4 V5
1  A  B  C  D  E
2  1  4  1  1  1
3  2  5  1  1  2
4  2  5  1  1  2
5  3  6  1  2  3</code></pre>
</div>
</div>
<p>You can also use the <code>View()</code> command to open data frames in the file pane.</p>
<p>Hmmm, something isn’t right with our rows here…</p>
<p>By default the <code>read.table</code> function assumes certain things from the file</p>
<ul>
<li>The file is a plain text file (there are separate functions to read excel files etc.)</li>
<li>Columns are separated by any number of tabs or spaces</li>
<li>There are the same number of data points in each column</li>
<li>There is <strong>no</strong> header row (labels for the columns)</li>
<li>There is no column with names for the rows</li>
</ul>
<p>If any of these are FALSE, then we need to give that information to the function. If it has a header column use the ‘header=TRUE’ argument.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/data.tsv"</span>, <span class="at">header=</span><span class="cn">TRUE</span>)  </span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that header=T will also work here. TRUE and FALSE can be abbreviated to T and F in R.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Note</strong> the comma between different parts of the functions arguments.</p>
<p>This overwrites our previous data frame <code>mydata</code> with the correctly formatted one. Let’s look at the first 5 rows again.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mydata, <span class="at">n=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  A B C D E
1 1 4 1 1 1
2 2 5 1 1 2
3 2 5 1 1 2
4 3 6 1 2 3
5 3 6 1 2 3</code></pre>
</div>
</div>
<p>Now we can see that the first row contains the column labels (A, B, C, D, E). R has automatically assigned default row names (1, 2, 3 etc.)</p>
<section id="other-options-for-read.table" class="level3">
<h3 class="anchored" data-anchor-id="other-options-for-read.table">Other options for read.table</h3>
<p>Use <code>sep =</code> to define how columns are separated in your input file. This file uses the tab character which we can write as “\t”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>mydata  <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/data.tsv"</span>, <span class="at">header =</span> T, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, <code>read.table</code> assumes columns are separated by any amount of white space (space or tabs). This can lead to problems if some of your columns have missing values, so it is good practice to always give a <code>sep</code> argument.</p>
<p>If your data has missing values you can use <code>fill = TRUE</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>mydata  <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/data.tsv"</span>, <span class="at">header =</span> T, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">fill =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This tells R to fill empty spaces in columns with the ‘NA’ character.</p>
<p>As this is such a common task there are functions identical to read.table but with different default settings. e.g.&nbsp;<code>read.delim</code> and <code>read.csv</code>. Check out the help pages for each of these.</p>
</section>
<section id="importing-datasets" class="level3">
<h3 class="anchored" data-anchor-id="importing-datasets">Importing Datasets</h3>
<p>In the <em>Environment</em> pane in RStudio there is a button called <strong>Import Dataset</strong>. This can make importing data much easier and calls the <code>read.*</code> set of functions for you. The command used will be displayed on the console. Make sure you save this code so that you can easily re-run the analysis later.</p>
<p>Note that you need to have the file on the computer to use this button.</p>
</section>
</section>
<section id="inspect-and-summarise-data" class="level2">
<h2 class="anchored" data-anchor-id="inspect-and-summarise-data">Inspect and summarise data</h2>
<p>Let’s use some simple functions to inspect and summarise our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mydata) <span class="co"># Shows the structure of a data frame and the data types of each column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   27 obs. of  5 variables:
 $ A: int  1 2 2 3 3 3 3 3 4 4 ...
 $ B: int  4 5 5 6 6 6 6 6 7 7 ...
 $ C: int  1 1 1 1 1 1 1 1 1 1 ...
 $ D: int  1 1 1 2 2 2 2 2 2 2 ...
 $ E: int  1 2 2 3 3 4 3 3 4 4 ...</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mydata) <span class="co"># Summarises the values in each column </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       A               B                C               D        
 Min.   :1.000   Min.   : 4.000   Min.   :1.000   Min.   :1.000  
 1st Qu.:3.000   1st Qu.: 6.000   1st Qu.:1.000   1st Qu.:2.000  
 Median :4.000   Median : 7.000   Median :1.000   Median :3.000  
 Mean   :4.296   Mean   : 7.296   Mean   :1.778   Mean   :3.333  
 3rd Qu.:5.000   3rd Qu.: 8.000   3rd Qu.:2.000   3rd Qu.:4.000  
 Max.   :8.000   Max.   :11.000   Max.   :5.000   Max.   :9.000  
       E        
 Min.   :1.000  
 1st Qu.:3.500  
 Median :4.000  
 Mean   :4.407  
 3rd Qu.:5.000  
 Max.   :8.000  </code></pre>
</div>
</div>
<p>Individual columns in a data frame can be accessed using the <code>$</code> sign (<code>mydata$A</code> <code>mydata$B</code> etc.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>mydata<span class="sc">$</span>A <span class="co"># Prints all values in column A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1 2 2 3 3 3 3 3 4 4 4 4 4 4 4 4 4 5 5 5 5 6 6 6 7 7 8</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mydata<span class="sc">$</span>A) <span class="co"># Summary information for column A only</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  1.000   3.000   4.000   4.296   5.000   8.000 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(mydata<span class="sc">$</span>A) <span class="co"># Mean of column A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4.296296</code></pre>
</div>
</div>
<p>Try some other summary functions like <code>mean</code>, <code>median</code>, <code>min</code> and <code>max</code>.</p>
</section>
<section id="square-bracket-notation" class="level2">
<h2 class="anchored" data-anchor-id="square-bracket-notation">Square bracket notation</h2>
<p>We can access specific rows, columns and cells within a data frame using square brackets: TABLE[ROW,COLUMN]. You can experiment with this notation by trying out the following commands.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Print the value in the first column of the first row</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>mydata[<span class="dv">1</span>,<span class="dv">1</span>]</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Use blanks to print an entire row or column</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>mydata[<span class="dv">2</span>,]</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>mydata[,<span class="dv">3</span>]</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="do">## You can select multiple rows and columns with ranges (:) or the c() function</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a><span class="do">## E.g. Print the first 5 rows of columns 3 and 5</span></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>mydata[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">5</span>)]</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a><span class="do">## You can also use row or column names instead of numbers</span></span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>mydata[,<span class="st">"B"</span>]</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a><span class="do">## You can select rows or columns based on certain criteria (subsetting).</span></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a><span class="do">## E.g. Select all rows where B is greater than 7</span></span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>mydata[mydata<span class="sc">$</span>B <span class="sc">&gt;</span> <span class="dv">7</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ordering-data" class="level2">
<h2 class="anchored" data-anchor-id="ordering-data">Ordering data</h2>
<p>The <code>order</code> function can be used to sort data frames by a specific column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">order</span>(mydata<span class="sc">$</span>B) <span class="co"># Returns the row numbers in the sorted order of column B</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
[26] 26 27</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use this in the row index position to sort the entire data frame by column B</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>mydata[<span class="fu">order</span>(mydata<span class="sc">$</span>B), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A  B C D E
1  1  4 1 1 1
2  2  5 1 1 2
3  2  5 1 1 2
4  3  6 1 2 3
5  3  6 1 2 3
6  3  6 1 2 4
7  3  6 1 2 3
8  3  6 1 2 3
9  4  7 1 2 4
10 4  7 1 2 4
11 4  7 1 2 4
12 4  7 1 2 5
13 4  7 1 2 5
14 4  7 1 3 4
15 4  7 1 3 4
16 4  7 1 3 4
17 4  7 2 3 4
18 5  8 2 3 5
19 5  8 2 4 5
20 5  8 2 4 5
21 5  8 2 4 5
22 6  9 3 5 6
23 6  9 3 5 6
24 6  9 3 6 6
25 7 10 4 7 7
26 7 10 4 8 7
27 8 11 5 9 8</code></pre>
</div>
</div>
<div class="challenge">
<h2 class="anchored">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<section id="see-if-you-can-do-the-following" class="level3">
<h3 class="anchored">See if you can do the following:</h3>
<ol type="1">
<li>Select the 11th value in the third column</li>
<li>Select all rows where D equals 4 (hint; use ‘==’)</li>
<li>Select rows where B has its maximum value (hint: use the max function)</li>
<li>Select even numbered rows only (hint: take a look at the seq function ‘?seq()’)</li>
<li>Select columns A, C and E</li>
<li>Sort table by decreasing order of column B (hint: look at the options in the order function)</li>
</ol>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored" data-anchor-id="see-if-you-can-do-the-following">
<i class="far fa-eye"></i> Solution:
</h2>
<ol type="1">
<li>Select row 11, column 3</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>mydata[<span class="dv">11</span>,<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Select rows where D 3 equals 4</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>mydata[mydata<span class="sc">$</span>D <span class="sc">==</span> <span class="dv">4</span>,] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A B C D E
19 5 8 2 4 5
20 5 8 2 4 5
21 5 8 2 4 5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Note the use of "==" for equality testing. This is a standard programming convention.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>Select rows where B has its maximum value</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>mydata[mydata<span class="sc">$</span>B <span class="sc">==</span> <span class="fu">max</span>(mydata<span class="sc">$</span>B), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A  B C D E
27 8 11 5 9 8</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Select even numbered rows only</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>mydata[<span class="fu">seq</span>(<span class="dv">2</span>,<span class="dv">26</span>, <span class="at">by =</span> <span class="dv">2</span>), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A  B C D E
2  2  5 1 1 2
4  3  6 1 2 3
6  3  6 1 2 4
8  3  6 1 2 3
10 4  7 1 2 4
12 4  7 1 2 5
14 4  7 1 3 4
16 4  7 1 3 4
18 5  8 2 3 5
20 5  8 2 4 5
22 6  9 3 5 6
24 6  9 3 6 6
26 7 10 4 8 7</code></pre>
</div>
</div>
<ol start="5" type="1">
<li>Select columns A, C and E</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>mydata[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>)] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A C E
1  1 1 1
2  2 1 2
3  2 1 2
4  3 1 3
5  3 1 3
6  3 1 4
7  3 1 3
8  3 1 3
9  4 1 4
10 4 1 4
11 4 1 4
12 4 1 5
13 4 1 5
14 4 1 4
15 4 1 4
16 4 1 4
17 4 2 4
18 5 2 5
19 5 2 5
20 5 2 5
21 5 2 5
22 6 3 6
23 6 3 6
24 6 3 6
25 7 4 7
26 7 4 7
27 8 5 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Or mydata[,c('A','C','E')] </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="6" type="1">
<li>Sort table by decreasing order of column B</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>mydata[<span class="fu">order</span>(mydata<span class="sc">$</span>B, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A  B C D E
27 8 11 5 9 8
25 7 10 4 7 7
26 7 10 4 8 7
22 6  9 3 5 6
23 6  9 3 5 6
24 6  9 3 6 6
18 5  8 2 3 5
19 5  8 2 4 5
20 5  8 2 4 5
21 5  8 2 4 5
9  4  7 1 2 4
10 4  7 1 2 4
11 4  7 1 2 4
12 4  7 1 2 5
13 4  7 1 2 5
14 4  7 1 3 4
15 4  7 1 3 4
16 4  7 1 3 4
17 4  7 2 3 4
4  3  6 1 2 3
5  3  6 1 2 3
6  3  6 1 2 4
7  3  6 1 2 3
8  3  6 1 2 3
2  2  5 1 1 2
3  2  5 1 1 2
1  1  4 1 1 1</code></pre>
</div>
</div>
</div>
</details>
</section>
</div>
<p><br></p>
</section>
<section id="filtering-data" class="level2">
<h2 class="anchored" data-anchor-id="filtering-data">Filtering data</h2>
<p>Square bracket notation can be handy for quickly filtering data frames or checking values, but can get cumbersome for more complex queries.</p>
<p>There is a <code>subset()</code> function in R specifically for filtering tables. This generally works better than using square brackets as it copes well with NA and NULL values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(mydata, mydata<span class="sc">$</span>C <span class="sc">==</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A B C D E
22 6 9 3 5 6
23 6 9 3 5 6
24 6 9 3 6 6</code></pre>
</div>
</div>
<p>The <code>tidyverse</code> packages have their own set of functions for filtering data and we will explore these in a later module.</p>
</section>
<section id="plotting-with-r" class="level2">
<h2 class="anchored" data-anchor-id="plotting-with-r">Plotting with R</h2>
<p>Base R has many built in functions for plotting data.</p>
<p>We recommend learning <code>ggplot2</code> for more complex graphics but it is useful to know what base R offers. Remember, to get more information about the options available to a function, type <code>?function</code>.</p>
<section id="histograms" class="level3">
<h3 class="anchored">Histograms</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(mydata<span class="sc">$</span>A)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The <code>ChickWeight</code> data set is another data frame built into R. It is a larger dataset, containing the weights of chicks on different diets, measured over time. Inspect the <code>ChickWeight</code> data using some of the functions you have used already.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="histograms">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<ol type="1">
<li>What type of data is stored in the <code>weight</code> column?</li>
<li>What is the mean weight of all the chicks?</li>
<li>What type of data is stored in the <code>Diet</code> column?</li>
<li>How many rows are in this table (try the <code>dim</code> or <code>nrow</code> functions)?</li>
</ol>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<ol type="1">
<li>What type of data is stored in the <code>weight</code> column?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(ChickWeight<span class="sc">$</span>weight)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>What is the mean weight of all the chicks?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(ChickWeight<span class="sc">$</span>weight)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 121.8183</code></pre>
</div>
</div>
<ol start="3" type="1">
<li>What type of data is stored in the <code>Diet</code> column?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(ChickWeight<span class="sc">$</span>Diet)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "factor"</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>How many rows are in this table (try the <code>dim</code> or <code>nrow</code> functions)?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(ChickWeight)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 578</code></pre>
</div>
</div>
</div>
</details>
</div>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(ChickWeight<span class="sc">$</span>weight)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can modify the number of vertical columns in a histogram with the argument <code>breaks</code>, to give us increased resolution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(ChickWeight<span class="sc">$</span>weight, <span class="at">breaks =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="boxplots" class="level3">
<h3 class="anchored" data-anchor-id="boxplots">Boxplots</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(mydata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(mydata<span class="sc">$</span>A, mydata<span class="sc">$</span>B, <span class="at">names=</span><span class="fu">c</span>(<span class="st">"Value A"</span>, <span class="st">"Value B"</span>) , <span class="at">ylab=</span><span class="st">"Count of Something"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="scatter-plots" class="level3">
<h3 class="anchored" data-anchor-id="scatter-plots">Scatter plots</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> mydata<span class="sc">$</span>A, <span class="at">y =</span> mydata<span class="sc">$</span>B)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/wd-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Take a look at <code>?plot</code>. You will see that the first argument is ‘x’ and the second is ‘y’. If you don’t name the arguments, the function will assume that the first argument is ‘x’ and the second is ‘y’.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This works because the first variable is assumed to be x and the second y</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mydata<span class="sc">$</span>A, mydata<span class="sc">$</span>B)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you name the arguments, you can put them in any order.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This works because the arguments are named</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">y =</span> mydata<span class="sc">$</span>B, <span class="at">x =</span> mydata<span class="sc">$</span>A)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Earlier, we used the plot function with a single x argument: <code>plot(cars)</code>. This worked because the x argument also accepts a data frame and will plot all combinations of columns. Try this for <code>mydata</code> and see what happens.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mydata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="saving-images" class="level2">
<h2 class="anchored" data-anchor-id="saving-images">Saving images</h2>
<p>There are a few ways to save images in RStudio:</p>
<section id="use-the-export-button-in-the-plots-pane-in-rstudio." class="level3">
<h3 class="anchored" data-anchor-id="use-the-export-button-in-the-plots-pane-in-rstudio.">Use the <strong>export</strong> button in the Plots pane in Rstudio.</h3>
<ul>
<li>Easy to perform.</li>
<li>Doesn’t work well for automated scripts.</li>
<li>No reproducible code to show how the image was generated.</li>
</ul>
</section>
<section id="use-a-graphics-device-function-in-your-r-code." class="level3">
<h3 class="anchored" data-anchor-id="use-a-graphics-device-function-in-your-r-code.">Use a <strong>graphics device</strong> function in your R code.</h3>
<p>Use the <code>png</code> function to save a png file (easy to load into web applications and presentations).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"filename.png"</span>) </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(A, B, <span class="at">names=</span><span class="fu">c</span>(<span class="st">"Value A"</span>, <span class="st">"Value B"</span>) , <span class="at">ylab=</span><span class="st">"Count of Something"</span>)</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>dev.off()</code> function closes the graphics device. In the code above, everything between <code>png()</code> and <code>dev.off()</code> is saved to ‘filename.png’.</p>
<p>You can also save as a pdf.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"filename.pdf"</span>) </span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(A, B, <span class="at">names=</span><span class="fu">c</span>(<span class="st">"Value A"</span>, <span class="st">"Value B"</span>) , <span class="at">ylab=</span><span class="st">"Count of Something"</span>)</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="statistical-testing" class="level2">
<h2 class="anchored" data-anchor-id="statistical-testing">Statistical testing</h2>
<p>R has many functions for statistical testing. We will cover a couple of examples here, but there are many more to explore. Remember, to get more information about the options available to a function, type <code>?function</code>.</p>
<section id="t-test" class="level3">
<h3 class="anchored">T-test</h3>
<p>The <code>t.test</code> function can be used to compare the <strong>numerical</strong> means of two groups. The t-test should only be used under the following assumptions.</p>
<ul>
<li>Each group is approximately normally distributed</li>
<li>The variance of the two groups is approximately equal</li>
<li>The two groups are independent of each other</li>
</ul>
<p>We can use R to simulate a data frame of values for two groups, WT and KO, with 10 values in each group (this could be any sort of experimental measurement).</p>
<p>Below, we use the <code>rnorm</code> function to generate random numbers from a normal distribution. The <code>mean</code> and <code>sd</code> arguments specify the mean and standard deviation of the distribution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># Set seed for reproducibility.</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Sample =</span> <span class="fu">paste</span>(<span class="st">"Sample"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">sep=</span><span class="st">"_"</span>),</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Group =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"KO"</span>), <span class="at">each=</span><span class="dv">10</span>),</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">Value =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">mean=</span><span class="dv">5</span>, <span class="at">sd=</span><span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">10</span>, <span class="at">mean=</span><span class="dv">6</span>, <span class="at">sd=</span><span class="dv">1</span>))</span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample Group    Value
1   Sample_1    WT 4.439524
2   Sample_2    WT 4.769823
3   Sample_3    WT 6.558708
4   Sample_4    WT 5.070508
5   Sample_5    WT 5.129288
6   Sample_6    WT 6.715065
7   Sample_7    WT 5.460916
8   Sample_8    WT 3.734939
9   Sample_9    WT 4.313147
10 Sample_10    WT 4.554338
11 Sample_11    KO 7.224082
12 Sample_12    KO 6.359814
13 Sample_13    KO 6.400771
14 Sample_14    KO 6.110683
15 Sample_15    KO 5.444159
16 Sample_16    KO 7.786913
17 Sample_17    KO 6.497850
18 Sample_18    KO 4.033383
19 Sample_19    KO 6.701356
20 Sample_20    KO 5.527209</code></pre>
</div>
</div>
<div class="discussion">
<h2 class="anchored" data-anchor-id="t-test">
<i class="far fa-bell"></i> Discussion
</h2>
<p>What is a normal distribution? How does <code>rnorm</code> work?</p>
<p>What does <code>set.seed</code> do?</p>
<p>What do the <code>paste</code> and <code>rep</code> functions do?</p>
</div>
<p><br></p>
<p>We can use the <code>t.test</code> function to compare the means of the two groups. The syntax for the t-test is <code>t.test(Value ~ Group, data=df)</code>, where <code>Value</code> is the numeric variable we want to compare and <code>Group</code> is the categorical variable that defines the groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(Value <span class="sc">~</span> Group, <span class="at">data =</span> df, <span class="at">var.equal =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  Value by Group
t = 2.5438, df = 18, p-value = 0.02036
alternative hypothesis: true difference in means between group KO and group WT is not equal to 0
95 percent confidence interval:
 0.1974232 2.0705694
sample estimates:
mean in group KO mean in group WT 
        6.208622         5.074626 </code></pre>
</div>
</div>
<div class="discussion">
<h2 class="anchored">
<i class="far fa-bell"></i> Discussion
</h2>
<p>How do we interpret the output of the t-test?</p>
<p>Are the mean values of the WT and KO groups significantly different?</p>
</div>
<p><br></p>
</section>
<section id="fishers-exact-test" class="level3">
<h3 class="anchored">Fisher’s exact test</h3>
<p>The <code>fisher.test</code> function can be used to test for association between two <strong>categorical</strong> variables.</p>
<section id="example" class="level4">
<h4 class="anchored">Example</h4>
<p>Researchers are testing a new anti-cancer drug in a mouse model. Tumour-bearing mice are randomly assigned to one of two <strong>treatment</strong> groups:</p>
<ul>
<li>Control (vehicle only)</li>
<li>Drug-treated</li>
</ul>
<p>After 4 weeks of treatment, tumours are assessed for <strong>response</strong>. Each mouse is classified as:</p>
<ul>
<li>Responder (tumour size reduced ≥ 30%)</li>
<li>Non-responder (tumour did not shrink sufficiently)</li>
</ul>
<p>Both variables are categorical:</p>
<ul>
<li>Treatment: Control / Drug</li>
<li>Response: Responder / Non-responder</li>
</ul>
<p>Let’s create a contingency table of the two categorical variables, <code>Treatment</code> and <code>Response</code>, and use the <code>fisher.test</code> function to test for an association between them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a contingency table</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>ct <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">17</span>,   <span class="co"># Control: Responder, Non-responder</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>    <span class="dv">12</span>, <span class="dv">8</span>),  <span class="co"># Drug: Responder, Non-responder</span></span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ct) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Control"</span>, <span class="st">"Drug"</span>)</span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ct) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Responder"</span>, <span class="st">"Non_responder"</span>)</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a>ct</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Responder Non_responder
Control         3            17
Drug           12             8</code></pre>
</div>
</div>
<p>Now we can perform the test on this contingency table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(ct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  ct
p-value = 0.007912
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 0.01758347 0.63708839
sample estimates:
odds ratio 
 0.1250459 </code></pre>
</div>
</div>
<div class="discussion">
<h2 class="anchored" data-anchor-id="example">
<i class="far fa-bell"></i> Discussion
</h2>
<p>Is there a significant association between treatment and response?</p>
</div>
<p><br></p>
</section>
</section>
</section>
<section id="case-study-quantitative-pcr" class="level2">
<h2 class="anchored" data-anchor-id="case-study-quantitative-pcr">Case Study: Quantitative PCR</h2>
<p>Quantitative PCR (qPCR) is a laboratory technique used to measure how much of a specific sequence is present in a sample. In gene expression experiments, RNA is first converted to complementary DNA (cDNA), and then qPCR is used to quantify how much of a particular gene is expressed.</p>
<p>Unlike traditional PCR, qPCR measures amplification in real time using fluorescent dyes or probes. As DNA is amplified during each cycle, fluorescence increases. The instrument monitors this fluorescence and determines when it rises above a defined background threshold.</p>
<p>The key measurement in qPCR is the <strong>Ct</strong> value (Cycle threshold). The Ct is the number of PCR cycles required for fluorescence to cross a detection threshold. It is inversely related to the amount of starting material.</p>
<section id="interpretation" class="level3">
<h3 class="anchored" data-anchor-id="interpretation">Interpretation</h3>
<ul>
<li>Low Ct → lots of starting material → high gene expression</li>
<li>High Ct → little starting material → low gene expression</li>
</ul>
<p>For example:</p>
<table class="table">
<thead>
<tr class="header">
<th>Sample</th>
<th>Ct value</th>
<th>Interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Sample A</td>
<td>18</td>
<td>High expression</td>
</tr>
<tr class="even">
<td>Sample B</td>
<td>25</td>
<td>Moderate expression</td>
</tr>
<tr class="odd">
<td>Sample C</td>
<td>32</td>
<td>Low expression</td>
</tr>
</tbody>
</table>
<p>Each PCR cycle approximately doubles the amount of DNA. That means:</p>
<ul>
<li>A difference of 1 Ct ≈ 2-fold difference in starting quantity</li>
<li>A difference of 3 Ct ≈ ~8-fold difference</li>
<li>A difference of 10 Ct ≈ ~1000-fold difference</li>
</ul>
</section>
<section id="reference-gene-normalisation" class="level3">
<h3 class="anchored" data-anchor-id="reference-gene-normalisation">Reference gene normalisation</h3>
<p>Biological samples vary in:</p>
<ul>
<li>Total RNA amount</li>
<li>RNA quality</li>
<li>Pipetting variation</li>
</ul>
<p>To control for this, we measure a reference (housekeeping) gene that should not change between conditions.</p>
<p>We then calculate ΔCt by subtracting the Ct of the reference gene from the Ct of the target gene for each sample.</p>
<p><span class="math display">\[
ΔCt = Ct_{target} - Ct_{reference}
\]</span></p>
</section>
<section id="comparing-conditions" class="level3">
<h3 class="anchored" data-anchor-id="comparing-conditions">Comparing conditions</h3>
<p>To compare gene expression between conditions, we calculate the difference between the ΔCt of the treatment group and the ΔCt of the control group, which is called ΔΔCt:</p>
<p><span class="math display">\[
ΔΔCt = ΔCt_{treatment} - ΔCt_{control}
\]</span></p>
<p>We can then calculate the fold change in expression using the following formula:</p>
<p><span class="math display">\[
Fold~Change = 2^{-ΔΔCt}
\]</span></p>
</section>
<section id="step-1-load-the-data" class="level3">
<h3 class="anchored">Step 1: Load the data</h3>
<p>Our qPCR data consists of Ct values for a gene of interest and a reference gene, across 16 samples, split into two conditions (Control and Treatment). We will use this data to compare gene expression between the two conditions.</p>
<p>You can download the data from the URL below and save it in a new folder in your working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create a folder in the working directory</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>)</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Download the qPCR file and save it in the data folder</span></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/training/DSB/data/qPCR_data.xlsx"</span>, <span class="at">destfile =</span> <span class="st">"data/qpcr_data.xlsx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Our data is in an Excel spreadsheet. We could open Excel and export the data as a tab delimited text file, then read it in using <code>read.table</code> or <code>read.delim</code>. However, R has a useful package called <code>readxl</code> for importing Excel files directly.</p>
<p>To access this package, we first need to install it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"readxl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now load the package and use the <em>readxl</em> functions in our current R session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="challenge">
<h2 class="anchored" data-anchor-id="step-1-load-the-data">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Use a function in <em>readxl</em> to read in the qPCR data from the Excel file you downloaded. You can use the help documentation to find out how to do this.</p>
<ol type="1">
<li>How many rows and columns are in the data frame?</li>
<li>What are the column names?</li>
<li>How many conditions are there and what are they called?</li>
<li>How many biological replicates do we have for each condition?</li>
<li>What is the mean Ct value for the gene of interest in the control samples?</li>
</ol>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>Read in the qpcr data and inspect it using the functions we have already seen.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>qpcr <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"data/qPCR_data.xlsx"</span>)</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [16 × 4] (S3: tbl_df/tbl/data.frame)
 $ Sample_ID: chr [1:16] "S1" "S2" "S3" "S4" ...
 $ Condition: chr [1:16] "Control" "Control" "Control" "Control" ...
 $ ct_gene  : num [1:16] 25.4 24.9 25.1 25.6 24.8 25.2 25 25.3 23.7 23.5 ...
 $ ct_ref   : num [1:16] 20.2 20.1 20.3 20 20.1 20.2 20 20.1 20.2 20 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample_ID          Condition            ct_gene          ct_ref     
 Length:16          Length:16          Min.   :23.30   Min.   :20.00  
 Class :character   Class :character   1st Qu.:23.57   1st Qu.:20.07  
 Mode  :character   Mode  :character   Median :24.35   Median :20.10  
                                       Mean   :24.38   Mean   :20.12  
                                       3rd Qu.:25.12   3rd Qu.:20.20  
                                       Max.   :25.60   Max.   :20.30  </code></pre>
</div>
</div>
<ol type="1">
<li>How many rows and columns are in the data frame?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 16  4</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>What are the column names?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Sample_ID" "Condition" "ct_gene"   "ct_ref"   </code></pre>
</div>
</div>
<ol start="3" type="1">
<li>How many conditions are there and what are they called?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="do">## The unique function gives the unique values in a vector</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(qpcr<span class="sc">$</span>Condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Control"   "Treatment"</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>How many biological replicates do we have for each condition?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Show a table of the number of entries for each condition</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(qpcr<span class="sc">$</span>Condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  Control Treatment 
        8         8 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Or convert Condition to a factor and use summary</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">as.factor</span>(qpcr<span class="sc">$</span>Condition))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Control Treatment 
        8         8 </code></pre>
</div>
</div>
<ol start="5" type="1">
<li>What is the mean Ct value for the gene of interest in the control samples?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(qpcr<span class="sc">$</span>ct_gene[qpcr<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Control"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25.1625</code></pre>
</div>
</div>
</div>
</details>
</div>
<p><br></p>
</section>
<section id="step-2-calculate-δct" class="level3">
<h3 class="anchored" data-anchor-id="step-2-calculate-δct">Step 2: Calculate ΔCt</h3>
<p>To calculate ΔCt, we need to subtract the Ct value of the reference gene from the Ct value of the gene of interest for each sample.</p>
<p>We can create new columns in a data frame using the <code>$</code> symbol.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="do">## This operation is performed on each row of the data frame, so we get a new column with the ΔCt values for each sample.</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>qpcr<span class="sc">$</span>delta_ct <span class="ot">&lt;-</span> qpcr<span class="sc">$</span>ct_gene <span class="sc">-</span> qpcr<span class="sc">$</span>ct_ref</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>qpcr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 16 × 5
   Sample_ID Condition ct_gene ct_ref delta_ct
   &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;
 1 S1        Control      25.4   20.2      5.2
 2 S2        Control      24.9   20.1      4.8
 3 S3        Control      25.1   20.3      4.8
 4 S4        Control      25.6   20        5.6
 5 S5        Control      24.8   20.1      4.7
 6 S6        Control      25.2   20.2      5  
 7 S7        Control      25     20        5  
 8 S8        Control      25.3   20.1      5.2
 9 S9        Treatment    23.7   20.2      3.5
10 S10       Treatment    23.5   20        3.5
11 S11       Treatment    23.8   20.1      3.7
12 S12       Treatment    23.4   20.2      3.2
13 S13       Treatment    23.6   20.1      3.5
14 S14       Treatment    23.9   20        3.9
15 S15       Treatment    23.3   20.2      3.1
16 S16       Treatment    23.5   20.1      3.4</code></pre>
</div>
</div>
<p>A new column <code>delta_ct</code> has been added to the data frame, which contains the ΔCt values for each sample.</p>
</section>
<section id="step-3-visualise-the-data" class="level3">
<h3 class="anchored" data-anchor-id="step-3-visualise-the-data">Step 3: Visualise the data</h3>
<p>We can use a boxplot to visualise the ΔCt values in the control and treatment samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr,</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Delta Ct"</span>,</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Gene expression by condition"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-IntroR_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can clearly see a difference between the two conditions, but is this difference statistically significant? We can use a t-test to find out.</p>
</section>
<section id="step-4-perform-a-t-test" class="level3">
<h3 class="anchored">Step 4: Perform a t-test</h3>
<p>A t-test is a statistical test used to compare the means of two groups. In our case, we want to compare the mean ΔCt values between the control and treatment groups.</p>
<p>The t-test is a valid test to use if the data in each group meets the following assumptions.</p>
<ul>
<li>The data in each group is approximately <strong>normally distributed</strong></li>
<li>The <strong>variance</strong> of the two groups is approximately equal</li>
<li>The two groups are <strong>independent</strong> of each other</li>
</ul>
<p>Later on, we will see how to check these assumptions and use alternative tests if they are not met. For now, we can assume that the data meets these assumptions:</p>
<ul>
<li>Box plots show approximately normal distribution</li>
<li>Variance looks approximately equal</li>
<li>The two groups are independent (different samples in each group)</li>
</ul>
<div class="challenge">
<h2 class="anchored" data-anchor-id="step-4-perform-a-t-test">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Use a t-test to decide if the mean values of Control and Treatment differ significantly.</p>
<ol type="1">
<li><p>Does the treatment have a significant effect on gene expression?</p></li>
<li><p>Does the treatment increase or decrease gene expression?</p></li>
</ol>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>Use a t-test to decide if the mean values of Control and Treatment differ more than would be expected by chance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr, <span class="at">var.equal=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  delta_ct by Condition
t = 11.391, df = 14, p-value = 1.821e-08
alternative hypothesis: true difference in means between group Control and group Treatment is not equal to 0
95 percent confidence interval:
 1.268289 1.856711
sample estimates:
  mean in group Control mean in group Treatment 
                 5.0375                  3.4750 </code></pre>
</div>
</div>
<ol type="1">
<li>Does the treatment have a significant effect on gene expression?</li>
</ol>
<p>The p-value is less than 0.05, so we can reject the null hypothesis and conclude that there is a significant difference in gene expression between the control and treatment groups.</p>
<p>You can save the result of the t-test as an object and print the p-value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">t.test</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr, <span class="at">var.equal=</span><span class="cn">TRUE</span>)</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>result<span class="sc">$</span>p.value</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.820703e-08</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Does the treatment increase or decrease gene expression?</li>
</ol>
<p>The mean ΔCt value for the treatment group is lower than the control group, which indicates that the treatment increases gene expression (remember that a lower ΔCt means higher expression).</p>
</div>
</details>
</div>
<p><br></p>
</section>
<section id="step-5-calculate-fold-change" class="level3">
<h3 class="anchored" data-anchor-id="step-5-calculate-fold-change">Step 5: Calculate fold change</h3>
<p>We now know that there is a significant difference in gene expression between the control and treatment groups, but how much of a difference is there? We can calculate the fold change in gene expression using the formula:</p>
<p><span class="math display">\[
Fold~Change = 2^{-ΔΔCt}
\]</span></p>
<p>To calculate ΔΔCt, we need to subtract the mean ΔCt of the control group from the mean ΔCt of the treatment group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>mean_control <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr<span class="sc">$</span>delta_ct[qpcr<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Control"</span>])</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>mean_treatment <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr<span class="sc">$</span>delta_ct[qpcr<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Treatment"</span>])</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>delta_delta_ct <span class="ot">&lt;-</span> mean_treatment <span class="sc">-</span> mean_control</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can calculate the fold change:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>fold_change <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>(<span class="sc">-</span>delta_delta_ct)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>fold_change</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.953652</code></pre>
</div>
</div>
<p>Expression is almost 3 times higher in the treatment group compared to the control group. When reporting this analysis, you should show the p-value and the fold change, as well as a visualisation of the data. The p-value tells us if the difference is statistically significant, while the fold change tells us how big the difference is (effect size).</p>
<p>We will revisit this case study in later modules to see how to do the same analysis using dedicated statistical and visualisation packages for high quality presentation.</p>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Congratulations! You have completed your first lesson in R. Hopefully you have a better understanding of what R is, its syntax and data types, and how to use it to import and analyse data.</p>
<p>Next we will start to look at the strengths of R as a programming language and how to use it to write scripts that can be easily reproduced and shared with others.</p>
<div class="key-points">
<h2 class="anchored">
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>R is a functional programming language for data science</li>
<li>RStudio is an interactive environment for programming in R</li>
<li>R has different data types and structures for organising data</li>
<li>R code is run within functions, which have different arguments for customisation</li>
<li>R functions can be used to import, manipulate and plot data</li>
<li>There are many functions for statistical analysis in R</li>
</ul>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>