<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Data Science for Biology - Programming in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/DRP_Logo_only_square_transparent.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Data Science for Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./01-start.html"> 
<span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r-an-rstudio" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">R an RStudio</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r-an-rstudio">    
        <li>
    <a class="dropdown-item" href="./02-Setup.html">
 <span class="dropdown-text">Course setup</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-IntroR.html">
 <span class="dropdown-text">Introduction to R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-Programming.html">
 <span class="dropdown-text">Programming in R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-HW.html">
 <span class="dropdown-text">Homework task</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#programming-in-r" id="toc-programming-in-r" class="nav-link active" data-scroll-target="#programming-in-r">Programming in R</a></li>
  <li><a href="#nested-functions" id="toc-nested-functions" class="nav-link" data-scroll-target="#nested-functions">Nested functions</a></li>
  <li><a href="#control-flow-statements" id="toc-control-flow-statements" class="nav-link" data-scroll-target="#control-flow-statements">Control flow statements</a>
  <ul class="collapse">
  <li><a href="#if-statements" id="toc-if-statements" class="nav-link" data-scroll-target="#if-statements">If statements</a></li>
  <li><a href="#for-loops" id="toc-for-loops" class="nav-link" data-scroll-target="#for-loops">For loops</a></li>
  </ul></li>
  <li><a href="#writing-functions" id="toc-writing-functions" class="nav-link" data-scroll-target="#writing-functions">Writing functions</a>
  <ul class="collapse">
  <li><a href="#custom-functions" id="toc-custom-functions" class="nav-link" data-scroll-target="#custom-functions">Custom functions</a></li>
  </ul></li>
  <li><a href="#reproducible-r-scripts" id="toc-reproducible-r-scripts" class="nav-link" data-scroll-target="#reproducible-r-scripts">Reproducible R scripts</a>
  <ul class="collapse">
  <li><a href="#running-r-scripts" id="toc-running-r-scripts" class="nav-link" data-scroll-target="#running-r-scripts">Running R scripts</a></li>
  <li><a href="#example-r-script" id="toc-example-r-script" class="nav-link" data-scroll-target="#example-r-script">Example R Script</a></li>
  <li><a href="#arguments" id="toc-arguments" class="nav-link" data-scroll-target="#arguments">Arguments</a></li>
  <li><a href="#commenting-and-formatting" id="toc-commenting-and-formatting" class="nav-link" data-scroll-target="#commenting-and-formatting">Commenting and formatting</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Programming in R</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>



<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>

<div class="objectives">
<h2 class="anchored">
<i class="far fa-check-square"></i> Learning Objectives
</h2>
<ul>
<li>Understand the power of R as a programming language</li>
<li>Learn to use control flow statements in R</li>
<li>Learn how to write your own functions</li>
<li>Write and run reproducible R scripts</li>
</ul>
</div>
<p><br></p>
<section id="programming-in-r" class="level2">
<h2 class="anchored" data-anchor-id="programming-in-r">Programming in R</h2>
<p>R is a powerful programming language that allows you to perform complex data analysis tasks.</p>
<p>One of the key advantages of using a programming language like R is that it allows you to automate repetitive tasks and publish code that is fully reproducible.</p>
<ul>
<li>Code can be shared with others, allowing them to reproduce your results</li>
<li>Code can be reused for different datasets or analyses</li>
<li>Code can be run in batch mode, allowing you to process large datasets or run multiple analyses at once</li>
</ul>
</section>
<section id="nested-functions" class="level2">
<h2 class="anchored" data-anchor-id="nested-functions">Nested functions</h2>
<p>R functions can be nested inside other functions. This allows you to perform multiple operations on data without creating intermediate variables. For example, you can use the <code>log</code> function inside the <code>mean</code> function to calculate the log of the mean of a set of numbers:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log</span>( <span class="fu">mean</span>(<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">7</span>)) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.481605</code></pre>
</div>
</div>
<p>R will perform the innermost function first, which is the <code>mean</code> function, and then pass the result to the <code>log</code> function.</p>
<p>We can also use nested functions to perform more complex operations. For example, we can use the <code>subset</code> function to filter the cars dataset for rows where the speed is greater than 10, and then use the <code>head</code> function to display the first 3 rows of the filtered dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( <span class="fu">subset</span>(cars, speed <span class="sc">&gt;</span> <span class="dv">10</span>), <span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   speed dist
10    11   17
11    11   28
12    12   14</code></pre>
</div>
</div>
<p>Nested functions are very powerful but can make your code difficult to understand if you nest many functions together. Later on we will learn how to use the R pipe operator <code>|&gt;</code> to combine functions in a more naturual way.</p>
</section>
<section id="control-flow-statements" class="level2">
<h2 class="anchored" data-anchor-id="control-flow-statements">Control flow statements</h2>
<p>Let’s begin by looking at some of the basic programming concepts in R. Control flow statements allow you to control the flow of your code based on certain conditions. These are essential for writing complex and dynamic code and common to most programming languages.</p>
<section id="if-statements" class="level3">
<h3 class="anchored">If statements</h3>
<p>If statements allow you to execute a block of code only if a certain condition is true. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;</span> <span class="dv">3</span>) {</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is greater than 3"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "x is greater than 3"</code></pre>
</div>
</div>
<p>If statements can also include an <strong>else</strong> clause, which allows you to execute a block of code if the condition is false:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;</span> <span class="dv">3</span>) {</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is greater than 3"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is not greater than 3"</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "x is not greater than 3"</code></pre>
</div>
</div>
<p>If statements can also be nested, allowing you to check multiple conditions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;</span> <span class="dv">3</span>) {</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is greater than 3"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (x <span class="sc">&lt;</span> <span class="dv">10</span>) { <span class="do">## This if statement only runs if the first condition is true</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"x is also less than 10"</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is not greater than 3"</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "x is greater than 3"
[1] "x is also less than 10"</code></pre>
</div>
</div>
<p>If statements can also be combined with logical operators to check multiple conditions at once:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;</span> <span class="dv">3</span> <span class="sc">&amp;&amp;</span> x <span class="sc">&lt;</span> <span class="dv">10</span>) {</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is greater than 3 and less than 10"</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "x is greater than 3 and less than 10"</code></pre>
</div>
</div>
<section id="other-logical-operators-in-r" class="level4">
<h4 class="anchored">Other logical operators in R</h4>
<ul>
<li><code>&amp;&amp;</code> - and</li>
<li><code>||</code> - or</li>
<li><code>!</code> - not</li>
<li><code>==</code> - equal to</li>
<li><code>!=</code> - not equal to</li>
<li><code>&gt;</code> - greater than</li>
<li><code>&lt;</code> - less than</li>
<li><code>&gt;=</code> - greater than or equal to</li>
<li><code>&lt;=</code> - less than or equal to</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">&gt;=</span> <span class="dv">5</span> <span class="sc">&amp;&amp;</span> y <span class="sc">!=</span> <span class="dv">15</span>) {</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is greater than or equal to 5 and y is not equal to 15"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "x is greater than or equal to 5 and y is not equal to 15"</code></pre>
</div>
</div>
<p>If statements and logic operators can be used with character strings.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="st">"banana"</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">==</span> <span class="st">"apple"</span>) {</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is an apple"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(x, <span class="st">"is not an apple"</span>))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "banana is not an apple"</code></pre>
</div>
</div>
<p>You can use <code>else if</code> to check multiple conditions in sequence.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="st">"pear"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">==</span> <span class="st">"apple"</span>) {</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is an apple"</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (x <span class="sc">==</span> <span class="st">"banana"</span>) {</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"x is a banana"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(x, <span class="st">"is not an apple or a banana"</span>))</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pear is not an apple or a banana"</code></pre>
</div>
</div>
<p>The next example has a logical value called <code>draw_plot</code> that is set to FALSE by default. The if statement checks if there are at least 5 data points in the <code>cars</code> dataset after filtering for rows with a speed less than <code>limit</code>. If there are, it sets <code>draw_plot</code> to TRUE and plots the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Only plot data if there are at least 5 data points</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>limit <span class="ot">=</span> <span class="dv">20</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>draw_plot <span class="ot">=</span> <span class="cn">FALSE</span> <span class="co"># Set plot to FALSE by default</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>cars_filtered <span class="ot">&lt;-</span> <span class="fu">subset</span>(cars,cars<span class="sc">$</span>speed <span class="sc">&lt;</span> limit)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(cars_filtered) <span class="sc">&gt;=</span> <span class="dv">5</span>) {</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  draw_plot <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (draw_plot <span class="sc">==</span> <span class="cn">TRUE</span>) {</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(cars)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-Programming_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Try setting limit to 7 and rerun the code. What happens?</p>
<p>If statements can also be used on vectors or data frames. For example, you can use an if statement to check if a certain value is present in a vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="dv">3</span> <span class="sc">%in%</span> x) {</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"3 is in the vector x"</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3 is in the vector x"</code></pre>
</div>
</div>
<p>The <code>%in%</code> is a special operator in R that is used to check if a value is present in a vector or data frame. It returns a logical value indicating if there is a match or not.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="other-logical-operators-in-r">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Write an if statement to determine if a treatment has a significant effect on gene expression based on a t-test result.</p>
<p>Use the following code to perform the t-test:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample data frame of gene expression values</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>qpcr_example <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Condition =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"Control"</span>, <span class="dv">5</span>), <span class="fu">rep</span>(<span class="st">"Treatment"</span>, <span class="dv">5</span>)),</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">delta_ct =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">3</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform t-test</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>qpcr_result <span class="ot">&lt;-</span> <span class="fu">t.test</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr_example, <span class="at">var.equal=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot a boxplot of the delta_ct values for the control and treatment groups. Add a title to the plot indicating whether the treatment has a significant effect on gene expression based on the t-test result.</p>
<p>See if you can add the p-value as a subtitle.</p>
<p><strong>HINTS</strong></p>
<ul>
<li>A title can be added with the <code>main</code> argument in the <code>boxplot</code> function.</li>
<li>You can use an if statement to determine the title based on the p-value from the t-test result.</li>
<li>Store the title text in an object and use that object in the <code>main</code> argument of the <code>boxplot</code> function.</li>
<li>The p-value can be accessed from the t-test result object with <code>qpcr_result$p.value</code>.</li>
<li>The subtitle can be added with the <code>sub</code> argument in the <code>boxplot</code> function.</li>
</ul>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the p-value from the t-test result</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (qpcr_result<span class="sc">$</span>p.value <span class="sc">&lt;=</span> <span class="fl">0.05</span>) {</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  title <span class="ot">&lt;-</span> <span class="st">"The treatment has a significant effect on gene expression."</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  title <span class="ot">&lt;-</span> <span class="st">"The treatment does not have a significant effect on gene expression."</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the boxplot</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr_example, <span class="at">main =</span> title, <span class="at">sub =</span> <span class="fu">paste</span>(<span class="st">"p-value:"</span>, <span class="fu">round</span>(qpcr_result<span class="sc">$</span>p.value, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-Programming_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</details>
</div>
<p><br></p>
</section>
</section>
<section id="for-loops" class="level3">
<h3 class="anchored">For loops</h3>
<p>For loops allow you to repeat a block of code a certain number of times. For example, you can use a for loop to iterate over a vector of values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Every time the loop runs, the variable i takes on the next value in the sequence from 1 to 5</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2
[1] 3
[1] 4
[1] 5
[1] 6</code></pre>
</div>
</div>
<p>For loops can also be used to iterate over the rows of a data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame of human patients and their ages</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">patient =</span> <span class="fu">c</span>(<span class="st">"Eve"</span>, <span class="st">"Arnold"</span>, <span class="st">"Fred"</span>),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">age =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">30</span>, <span class="dv">35</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  patient age
1     Eve  10
2  Arnold  30
3    Fred  35</code></pre>
</div>
</div>
<p>We can use a for loop to iterate over the rows of the data frame and print out a message for each patient:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use a for loop to iterate over the rows of the data frame</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(df)) {</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Use an if statement to classify patients as "Child" or "Adult" based on their age</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (df<span class="sc">$</span>age[i] <span class="sc">&lt;</span> <span class="dv">18</span>) {</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>     classification <span class="ot">&lt;-</span> <span class="st">"Child"</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span>{</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>     classification <span class="ot">&lt;-</span> <span class="st">"Adult"</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Print a statement for each patient</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(df<span class="sc">$</span>patient[i], <span class="st">"is visiting the doctor and is classified as"</span>, classification))</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Eve is visiting the doctor and is classified as Child"
[1] "Arnold is visiting the doctor and is classified as Adult"
[1] "Fred is visiting the doctor and is classified as Adult"</code></pre>
</div>
</div>
<div class="challenge">
<h2 class="anchored" data-anchor-id="for-loops">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Write code with control flow that takes a data frame of gene expression values and creates a new column <strong>expression</strong> classified as “low”, “medium”, or “high” based on the following criteria:</p>
<ul>
<li><strong>low</strong> if the expression value is less than 10</li>
<li><strong>medium</strong> if the expression value is between 10 and 100</li>
<li><strong>high</strong> if the expression value is greater than 100</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a sample data frame of gene expression values</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>gene_expression <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> <span class="fu">c</span>(<span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"gene3"</span>, <span class="st">"gene4"</span>, <span class="st">"gene5"</span>),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression_value =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">15</span>, <span class="dv">50</span>, <span class="dv">150</span>, <span class="dv">200</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new column to store the classifications, set to NA by default</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>gene_expression<span class="sc">$</span>expression <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>HINTS</strong></p>
<ul>
<li>You can use a <code>for</code> loop to iterate over the rows of the data frame</li>
<li>Use <code>if</code> <code>else if</code> and <code>else</code> statements to classify the expression values based on the criteria provided.</li>
</ul>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="do">## loop through the rows</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(gene_expression)) {</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Check the expression level and assign with if statements</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (gene_expression<span class="sc">$</span>expression_value[i] <span class="sc">&lt;</span> <span class="dv">10</span>) {</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    gene_expression<span class="sc">$</span>expression[i] <span class="ot">&lt;-</span> <span class="st">"low"</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (gene_expression<span class="sc">$</span>expression_value[i] <span class="sc">&gt;=</span> <span class="dv">10</span> <span class="sc">&amp;&amp;</span> gene_expression<span class="sc">$</span>expression_value[i] <span class="sc">&lt;=</span> <span class="dv">100</span>) {</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>    gene_expression<span class="sc">$</span>expression[i] <span class="ot">&lt;-</span> <span class="st">"medium"</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>    gene_expression<span class="sc">$</span>expression[i] <span class="ot">&lt;-</span> <span class="st">"high"</span></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="co"># View the resulting data frame</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_expression)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   gene expression_value expression
1 gene1                5        low
2 gene2               15     medium
3 gene3               50     medium
4 gene4              150       high
5 gene5              200       high</code></pre>
</div>
</div>
</div>
</details>
</div>
<p><br></p>
<p><strong>For loops</strong> and <strong>if statements</strong> are used in almost every programming language. They allow you to write code that can adapt to different situations and perform complex tasks.</p>
<p>However, they can be slow to run on large datasets and there are more efficient ways to manipulate data frames in R. We will cover some of these in later modules.</p>
</section>
</section>
<section id="writing-functions" class="level2">
<h2 class="anchored" data-anchor-id="writing-functions">Writing functions</h2>
<p>Functions are reusable blocks of code that perform a specific task. They allow you to break down complex problems into smaller, more manageable pieces.</p>
<p>You have already used many functions in R, such as <code>plot()</code>, <code>read.table()</code>, and <code>t.test()</code>. All of these functions have underlying code that performs a specific task when you call the function. In fact, you can print the code for any function in R by simply typing the name of the function without parentheses.</p>
<p>Try this with the <code>read.table</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>read.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>function (file, header = FALSE, sep = "", quote = "\"'", dec = ".", 
    numerals = c("allow.loss", "warn.loss", "no.loss"), row.names, 
    col.names, as.is = !stringsAsFactors, tryLogical = TRUE, 
    na.strings = "NA", colClasses = NA, nrows = -1, skip = 0, 
    check.names = TRUE, fill = !blank.lines.skip, strip.white = FALSE, 
    blank.lines.skip = TRUE, comment.char = "#", allowEscapes = FALSE, 
    flush = FALSE, stringsAsFactors = FALSE, fileEncoding = "", 
    encoding = "unknown", text, skipNul = FALSE) 
{
    if (missing(file) &amp;&amp; !missing(text)) {
        file &lt;- textConnection(text, encoding = "UTF-8")
        encoding &lt;- "UTF-8"
        on.exit(close(file))
    }
    if (is.character(file)) {
        file &lt;- if (nzchar(fileEncoding)) 
            file(file, "rt", encoding = fileEncoding)
        else file(file, "rt")
        on.exit(close(file))
    }
    if (!inherits(file, "connection")) 
        stop("'file' must be a character string or connection")
    if (!isOpen(file, "rt")) {
        open(file, "rt")
        on.exit(close(file))
    }
    pbEncoding &lt;- if (encoding %in% c("", "bytes", "UTF-8")) 
        encoding
    else "bytes"
    numerals &lt;- match.arg(numerals)
    if (skip &gt; 0L) 
        readLines(file, skip)
    nlines &lt;- n0lines &lt;- if (nrows &lt; 0L) 
        5
    else min(5L, (header + nrows))
    lines &lt;- .External(C_readtablehead, file, nlines, comment.char, 
        blank.lines.skip, quote, sep, skipNul)
    if (encoding %in% c("UTF-8", "latin1")) 
        Encoding(lines) &lt;- encoding
    nlines &lt;- length(lines)
    if (!nlines) {
        if (missing(col.names)) 
            stop("no lines available in input")
        rlabp &lt;- FALSE
        cols &lt;- length(col.names)
    }
    else {
        if (all(!nzchar(lines))) 
            stop("empty beginning of file")
        if (nlines &lt; n0lines &amp;&amp; file == 0L) {
            pushBack(c(lines, lines, ""), file, encoding = pbEncoding)
            on.exit((clearPushBack(stdin())))
        }
        else pushBack(c(lines, lines), file, encoding = pbEncoding)
        first &lt;- scan(file, what = "", sep = sep, quote = quote, 
            nlines = 1, quiet = TRUE, skip = 0, strip.white = TRUE, 
            blank.lines.skip = blank.lines.skip, na.strings = character(0), 
            comment.char = comment.char, allowEscapes = allowEscapes, 
            encoding = encoding, skipNul = skipNul)
        col1 &lt;- if (missing(col.names)) 
            length(first)
        else length(col.names)
        col &lt;- numeric(nlines - 1L)
        if (nlines &gt; 1L) 
            for (i in seq_along(col)) col[i] &lt;- length(scan(file, 
                what = "", sep = sep, quote = quote, nlines = 1, 
                quiet = TRUE, skip = 0, strip.white = strip.white, 
                blank.lines.skip = blank.lines.skip, comment.char = comment.char, 
                allowEscapes = allowEscapes, encoding = encoding, 
                skipNul = skipNul))
        cols &lt;- max(col1, col)
        rlabp &lt;- (cols - col1) == 1L
        if (rlabp &amp;&amp; missing(header)) 
            header &lt;- TRUE
        if (!header) 
            rlabp &lt;- FALSE
        if (header) {
            .External(C_readtablehead, file, 1L, comment.char, 
                blank.lines.skip, quote, sep, skipNul)
            if (missing(col.names)) 
                col.names &lt;- first
            else if (length(first) != length(col.names)) 
                warning("header and 'col.names' are of different lengths")
        }
        else if (missing(col.names)) 
            col.names &lt;- paste0("V", 1L:cols)
        if (length(col.names) + rlabp &lt; cols) 
            stop("more columns than column names")
        if (fill &amp;&amp; length(col.names) &gt; cols) 
            cols &lt;- length(col.names)
        if (!fill &amp;&amp; cols &gt; 0L &amp;&amp; length(col.names) &gt; cols) 
            stop("more column names than columns")
        if (cols == 0L) 
            stop("first five rows are empty: giving up")
    }
    if (check.names) 
        col.names &lt;- make.names(col.names, unique = TRUE)
    if (rlabp) 
        col.names &lt;- c("row.names", col.names)
    nmColClasses &lt;- names(colClasses)
    if (is.null(nmColClasses)) {
        if (length(colClasses) &lt; cols) 
            colClasses &lt;- rep_len(colClasses, cols)
    }
    else {
        tmp &lt;- rep_len(NA_character_, cols)
        names(tmp) &lt;- col.names
        i &lt;- match(nmColClasses, col.names, 0L)
        if (any(i &lt;= 0L)) 
            warning("not all columns named in 'colClasses' exist")
        tmp[i[i &gt; 0L]] &lt;- colClasses[i &gt; 0L]
        colClasses &lt;- tmp
    }
    what &lt;- rep.int(list(""), cols)
    names(what) &lt;- col.names
    colClasses[colClasses %in% c("real", "double")] &lt;- "numeric"
    known &lt;- colClasses %in% c("logical", "integer", "numeric", 
        "complex", "character", "raw")
    what[known] &lt;- lapply(colClasses[known], do.call, list(0))
    what[colClasses %in% "NULL"] &lt;- list(NULL)
    keep &lt;- !vapply(what, is.null, NA)
    data &lt;- scan(file = file, what = what, sep = sep, quote = quote, 
        dec = dec, nmax = nrows, skip = 0, na.strings = na.strings, 
        quiet = TRUE, fill = fill, strip.white = strip.white, 
        blank.lines.skip = blank.lines.skip, multi.line = FALSE, 
        comment.char = comment.char, allowEscapes = allowEscapes, 
        flush = flush, encoding = encoding, skipNul = skipNul)
    nlines &lt;- length(data[[which.max(keep)]])
    if (cols != length(data)) {
        warning("cols = ", cols, " != length(data) = ", length(data), 
            domain = NA)
        cols &lt;- length(data)
    }
    if (is.logical(as.is)) {
        as.is &lt;- rep_len(as.is, cols)
    }
    else if (is.numeric(as.is)) {
        if (any(as.is &lt; 1 | as.is &gt; cols)) 
            stop("invalid numeric 'as.is' expression")
        i &lt;- rep.int(FALSE, cols)
        i[as.is] &lt;- TRUE
        as.is &lt;- i
    }
    else if (is.character(as.is)) {
        i &lt;- match(as.is, col.names, 0L)
        if (any(i &lt;= 0L)) 
            warning("not all columns named in 'as.is' exist")
        i &lt;- i[i &gt; 0L]
        as.is &lt;- rep.int(FALSE, cols)
        as.is[i] &lt;- TRUE
    }
    else if (length(as.is) != cols) 
        stop(gettextf("'as.is' has the wrong length %d  != cols = %d", 
            length(as.is), cols), domain = NA)
    do &lt;- keep &amp; !known
    if (rlabp) 
        do[1L] &lt;- FALSE
    for (i in (1L:cols)[do]) {
        data[[i]] &lt;- if (is.na(colClasses[i])) 
            type.convert(data[[i]], as.is = as.is[i], dec = dec, 
                numerals = numerals, na.strings = character(0L), 
                tryLogical = tryLogical)
        else if (colClasses[i] == "factor") 
            as.factor(data[[i]])
        else if (colClasses[i] == "Date") 
            as.Date(data[[i]])
        else if (colClasses[i] == "POSIXct") 
            as.POSIXct(data[[i]])
        else methods::as(data[[i]], colClasses[i])
    }
    compactRN &lt;- TRUE
    if (missing(row.names)) {
        if (rlabp) {
            row.names &lt;- data[[1L]]
            data &lt;- data[-1L]
            keep &lt;- keep[-1L]
            compactRN &lt;- FALSE
        }
        else row.names &lt;- .set_row_names(as.integer(nlines))
    }
    else if (is.null(row.names)) {
        row.names &lt;- .set_row_names(as.integer(nlines))
    }
    else if (is.character(row.names)) {
        compactRN &lt;- FALSE
        if (length(row.names) == 1L) {
            rowvar &lt;- (1L:cols)[match(col.names, row.names, 0L) == 
                1L]
            row.names &lt;- data[[rowvar]]
            data &lt;- data[-rowvar]
            keep &lt;- keep[-rowvar]
        }
    }
    else if (is.numeric(row.names) &amp;&amp; length(row.names) == 1L) {
        compactRN &lt;- FALSE
        rlabp &lt;- row.names
        row.names &lt;- data[[rlabp]]
        data &lt;- data[-rlabp]
        keep &lt;- keep[-rlabp]
    }
    else stop("invalid 'row.names' specification")
    data &lt;- data[keep]
    if (is.object(row.names) || !(is.integer(row.names))) 
        row.names &lt;- as.character(row.names)
    if (!compactRN) {
        if (length(row.names) != nlines) 
            stop("invalid 'row.names' length")
        if (anyDuplicated(row.names)) 
            stop("duplicate 'row.names' are not allowed")
        if (anyNA(row.names)) 
            stop("missing values in 'row.names' are not allowed")
    }
    class(data) &lt;- "data.frame"
    attr(data, "row.names") &lt;- row.names
    data
}
&lt;bytecode: 0x15e011eb0&gt;
&lt;environment: namespace:utils&gt;</code></pre>
</div>
</div>
<p>To start with, the code is quite long. This should really highlight the power of functions in R. Instead of having to write all of this code every time you want to read in a table, you can simply call the <code>read.table</code> function and it will execute all of that code for you.</p>
<p>If you scroll to the beginning, you can see that the code begins with a call of <code>function</code>, which indicates that this is a function definition. This is followed by the arguments that the function takes, which are listed in parentheses <code>()</code>. In this case, the <code>read.table</code> function takes several arguments, such as <code>file</code>, <code>header</code>, and <code>sep</code>.</p>
<p>The code inside the curly braces <code>{}</code> is the body of the function, which contains the code and control flow statements that will be executed when the function is called.</p>
<section id="custom-functions" class="level3">
<h3 class="anchored">Custom functions</h3>
<p>R allows you to create your own custom functions using the same syntax. To create a function in R, you use the <code>function</code> keyword. For example, here is a simple function that takes two numbers as input, calculates the sum and outputs a log transformation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>log_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y) {</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  sum <span class="ot">&lt;-</span> <span class="fu">sum</span>(x,y)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  logsum <span class="ot">&lt;-</span> <span class="fu">log</span>(sum)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(logsum)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this example, we define a function called <code>log_sum</code> that takes two arguments, <code>x</code> and <code>y</code>. The body of the function calculates the <code>sum</code> of <code>x</code> and <code>y</code> and then transforms it using the <code>log</code> function.</p>
<p>The <code>return()</code> function is used to return the value of <code>logsum</code> when the function is called. If a return statement is not included, the function will just return the value of the last expression.</p>
<p>You can call this function by passing two numbers as arguments:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log_sum</span>(<span class="dv">3</span>, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.079442</code></pre>
</div>
</div>
<p>If we need to perform this calculation multiple times with different numbers, we can simply call the <code>log_sum</code> function with different arguments instead of writing the same code over and over again.</p>
<p>We can add default values to the function arguments. This allows us to call the function without providing values for those arguments, and the default values will be used instead. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>log_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">base =</span> <span class="fu">exp</span>(<span class="dv">1</span>)) {</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  sum <span class="ot">&lt;-</span> <span class="fu">sum</span>(x,y)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  logsum <span class="ot">&lt;-</span> <span class="fu">log</span>(sum, <span class="at">base =</span> base)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(logsum)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this example, we added a new argument called <code>base</code> with a default value of <code>exp(1)</code>, which is the natural logarithm base. If we call the function without providing a value for <code>base</code>, it will use the default value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log_sum</span>(<span class="dv">3</span>, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.079442</code></pre>
</div>
</div>
<p>If we want to use a different base for the logarithm, we can provide a value for the <code>base</code> argument when calling the function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log_sum</span>(<span class="dv">3</span>, <span class="dv">5</span>, <span class="at">base =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.90309</code></pre>
</div>
</div>
<p>Another useful feature of functions in R is that they can return multiple values as a list. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>log_sum <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">base =</span> <span class="fu">exp</span>(<span class="dv">1</span>)) {</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  sum <span class="ot">&lt;-</span> x <span class="sc">+</span> y</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  logsum <span class="ot">&lt;-</span> <span class="fu">log</span>(sum, <span class="at">base =</span> base)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">sum =</span> sum, <span class="at">logsum =</span> logsum))</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this example, the <code>log_sum</code> function returns a list containing both the sum and the log of the sum. You can access these values using the <code>$</code> operator:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">log_sum</span>(<span class="dv">3</span>, <span class="dv">5</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>result<span class="sc">$</span>sum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>result<span class="sc">$</span>logsum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.079442</code></pre>
</div>
</div>
<p>Functions are a fundamental part of programming in R. They allow you to write reusable code that can be easily maintained and shared with others. They also help to make your code more organised and easier to read.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="custom-functions">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Write a function called <code>qPCR_analysis</code> that performs all of the steps we used in the first session.</p>
<ol type="1">
<li>Load the data from “data/qpcr_results.xlsx” into a data frame.</li>
<li>Calculate the delta_ct values for each sample.</li>
<li>Perform a t-test to compare the delta_ct values between the two conditions.</li>
<li>Calculate the fold change in gene expression between the treatment and control groups.</li>
<li>Plot a boxplot of the delta_ct values for the control and treatment groups.</li>
</ol>
<p>The function should take a data frame as an argument and return a list containing the t-test result and the fold change. It should also plot a boxplot of delta_ct values per condition.</p>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>qPCR_analysis <span class="ot">&lt;-</span> <span class="cf">function</span>(qpcr_table) {</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate delta_ct values</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  qpcr_table<span class="sc">$</span>delta_ct <span class="ot">&lt;-</span> qpcr_table<span class="sc">$</span>ct_gene <span class="sc">-</span> qpcr_table<span class="sc">$</span>ct_ref</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot boxplot in base R</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">boxplot</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr_table, <span class="at">main =</span> <span class="st">"Delta Ct values by condition"</span>, <span class="at">ylab =</span> <span class="st">"Delta Ct"</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform t-test</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>  t_test_result <span class="ot">&lt;-</span> <span class="fu">t.test</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr_table, <span class="at">var.equal=</span><span class="cn">TRUE</span>)</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate fold change</span></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>  mean_control <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr_table<span class="sc">$</span>delta_ct[qpcr_table<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Control"</span>])</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>  mean_treatment <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr_table<span class="sc">$</span>delta_ct[qpcr_table<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Treatment"</span>])</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>  delta_delta_ct <span class="ot">&lt;-</span> mean_treatment <span class="sc">-</span> mean_control</span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>  fold_change <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>(<span class="sc">-</span>delta_delta_ct)</span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">t_test_result =</span> t_test_result, <span class="at">fold_change =</span> fold_change, <span class="at">boxplot =</span> boxplot))</span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>
</div>
<p><br></p>
<p>You can call this function by passing a data frame containing the qPCR results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>qpcr_result <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"data/qpcr_data.xlsx"</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the qPCR_analysis function</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">qPCR_analysis</span>(qpcr_result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-Programming_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the p-value and fold change</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"The p-value is "</span>, results<span class="sc">$</span>t_test_result<span class="sc">$</span>p.value, <span class="st">" and the fold change is "</span>, results<span class="sc">$</span>fold_change)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "The p-value is  1.82070315127758e-08  and the fold change is  2.953652291879"</code></pre>
</div>
</div>
<p>We have now written a function that performs all of the steps of our qPCR analysis. This function can be reused for different datasets by simply passing a different data frame as an argument.</p>
<p>Let’s load a second qPCR dataset to test our function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the second dataset</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/training/DSB/data/qPCR_data2.xlsx"</span>, <span class="at">destfile =</span> <span class="st">"data/qpcr_data2.xlsx"</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>qpcr_table2 <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"data/qpcr_data2.xlsx"</span>)</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the qPCR_analysis function</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>results2 <span class="ot">&lt;-</span> <span class="fu">qPCR_analysis</span>(qpcr_table2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="02-Programming_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(<span class="st">"The p-value is "</span>, results2<span class="sc">$</span>t_test_result<span class="sc">$</span>p.value, <span class="st">" and the fold change is "</span>, results2<span class="sc">$</span>fold_change)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "The p-value is  0.207646591718013  and the fold change is  0.726986258660155"</code></pre>
</div>
</div>
<p>How easy was that! This is the power of functions in R.</p>
</section>
</section>
<section id="reproducible-r-scripts" class="level2">
<h2 class="anchored" data-anchor-id="reproducible-r-scripts">Reproducible R scripts</h2>
<p>R scripts are plain text files that contain R code. They can be created and edited in any text editor, but using an environment like RStudio makes it easier to write and run your code.</p>
<p>Saving your code in an R script allows you to keep a record of your analysis and makes it easier to reproduce your results. You can also share your R scripts with others, allowing them to run the same code on different data.</p>
<section id="running-r-scripts" class="level3">
<h3 class="anchored" data-anchor-id="running-r-scripts">Running R scripts</h3>
<p>R scripts can be run interactively in RStudio by clicking the “Source” button or by using the <code>source()</code> function in the console. For example, if you have an R script called <em>analysis.R</em> in your working directory, you can run it with the following command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Example, don't run this</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"analysis.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R scripts can also be run from the Windows, Mac or Linux command line. This is useful for several reasons:</p>
<ul>
<li>Run R scripts without opening RStudio</li>
<li>Run R scripts with arguments (see below)</li>
<li>Include R scripts in larger workflows or pipelines</li>
<li>Run R scripts on a remote server or cluster</li>
<li>Run multiple R analyses at once</li>
</ul>
<p>To run an R script from the command line, you need to open a Terminal window. Most computers have a Terminal app installed. RStudio has an inbuilt Terminal, you can find it next to the Console tab.</p>
<ul>
<li><strong>Console</strong> is where you can run R commands interactively.</li>
<li><strong>Terminal</strong> is where you can run command line commands on your computer or server.</li>
</ul>
<p>To run an R script on the command line, use the <code>Rscript</code> command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Example, don't run this </span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> analysis.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="example-r-script" class="level3">
<h3 class="anchored">Example R Script</h3>
<p>Download the R script below and open it in RStudio for reading.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"scripts"</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"http://bifx-core3.bio.ed.ac.uk/training/DSB/scripts/qPCR_analysis.R"</span>, <span class="at">destfile =</span> <span class="st">"scripts/qpcr_analysis.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="discussion">
<h2 class="anchored" data-anchor-id="example-r-script">
<i class="fas fa-comments"></i> Discussion:
</h2>
<ol type="1">
<li>What does the script do?</li>
<li>What are the inputs and outputs of the script?</li>
<li>Where are the output files saved?</li>
<li>What are the advantages of using an R script instead of running code interactively in the console?</li>
</ol>
</div>
<p><br></p>
<p>Try running this script in RStudio and from the command line. Make sure you are in your working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> scripts/qpcr_analysis.R</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>What happens when you run the R Script?</li>
<li>Where does the output appear?</li>
</ul>
</section>
<section id="arguments" class="level3">
<h3 class="anchored">Arguments</h3>
<p>At the moment our script is not very flexible. It is hard coded to read in a specific file and save the output to a specific location.</p>
<p>We can make our script more flexible by using <strong>arguments</strong>. Arguments allow us to specify the input and output files on the command line instead of in the script. This way, we can change them easily without ever having to edit the code e.g.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">## These are just examples, don't run them yet.</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> qPCR_analysis.R data/qpcr_results.xlsx qpcr_plot_exp1.png</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="ex">OR</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> qPCR_analysis.R data/qpcr_results2.xlsx qpcr_plot_exp2.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R interprets everything after the script name as an argument to the script. In the example above, <code>data/qpcr_results.xlsx</code> is the input file containing the qPCR results and <code>qpcr_plot.png</code> is the name of the output file where the plot will be saved.</p>
<p>We need to make a few changes to our script for R to accept these argument. We can use the built-in function <code>commandArgs()</code> to access the arguments passed to the script from the command line. You can use this function to read in the input and output file paths as arguments. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly =</span> <span class="cn">TRUE</span>) <span class="do">## trailingOnly = TRUE means that only the arguments after the script name will be returned</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>input_file <span class="ot">&lt;-</span> args[<span class="dv">1</span>] <span class="co"># The first argument will be the input file path</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>output_file <span class="ot">&lt;-</span> args[<span class="dv">2</span>] <span class="co"># The second argument will be the output file path</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code will read in the first and second arguments passed to the script and store them in the <code>input_file</code> and <code>output_file</code> variables, respectively. You can then use these variables in your code to read in the data and save the plot.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="arguments">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>Modify the <code>qPCR_analysis.R</code> script to use the input and output file paths as arguments.</p>
<p><strong>HINTS</strong></p>
<ul>
<li>Use the <code>commandArgs()</code> function to read in the arguments from the command line.</li>
<li>Replace the hardcoded input file name with the <code>input_file</code> variable.</li>
<li>Replace the hardcoded output file name with the <code>output_file</code> variable to save the plot.</li>
</ul>
<details>
<summary>
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly =</span> <span class="cn">TRUE</span>) <span class="do">## trailingOnly = TRUE means that only the arguments after the script name will be returned</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>input_file <span class="ot">&lt;-</span> args[<span class="dv">1</span>]</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>output_file <span class="ot">&lt;-</span> args[<span class="dv">2</span>]</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>qpcr_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(input_file)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to perform qPCR analysis</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>qPCR_analysis <span class="ot">&lt;-</span> <span class="cf">function</span>(qpcr_table,plot_file) {</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate delta_ct values</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>  qpcr_table<span class="sc">$</span>delta_ct <span class="ot">&lt;-</span> qpcr_table<span class="sc">$</span>ct_gene <span class="sc">-</span> qpcr_table<span class="sc">$</span>ct_ref</span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform t-test</span></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a>  t_test_result <span class="ot">&lt;-</span> <span class="fu">t.test</span>(delta_ct <span class="sc">~</span> Condition, <span class="at">data =</span> qpcr_table, <span class="at">var.equal=</span><span class="cn">TRUE</span>)</span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate fold change</span></span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a>  mean_control <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr_table<span class="sc">$</span>delta_ct[qpcr_table<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Control"</span>])</span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a>  mean_treatment <span class="ot">&lt;-</span> <span class="fu">mean</span>(qpcr_table<span class="sc">$</span>delta_ct[qpcr_table<span class="sc">$</span>Condition <span class="sc">==</span> <span class="st">"Treatment"</span>])</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a>  delta_delta_ct <span class="ot">&lt;-</span> mean_treatment <span class="sc">-</span> mean_control</span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a>  fold_change <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>(<span class="sc">-</span>delta_delta_ct)</span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save boxplot as a PDF</span></span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">png</span>(plot_file)</span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">boxplot</span>(delta_ct <span class="sc">~</span> Condition, </span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a>          <span class="at">data =</span> qpcr_table, </span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a>          <span class="at">main =</span> <span class="st">"Delta Ct values by condition"</span>, </span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a>          <span class="at">ylab =</span> <span class="st">"Delta Ct"</span>,</span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a>          <span class="at">sub =</span> <span class="fu">paste</span>(<span class="st">"Fold change:"</span>, <span class="fu">round</span>(fold_change, <span class="dv">2</span>), </span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"p-value:"</span>, <span class="fu">round</span>(t_test_result<span class="sc">$</span>p.value, <span class="dv">4</span>))</span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dev.off</span>()</span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a><span class="fu">qPCR_analysis</span>(qpcr_data, output_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>
</div>
<p><br></p>
<p>Try running the modified script from the <em>terminal</em> window specifying the input and output file names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> scripts/qPCR_analysis.R data/qpcr_data.xlsx qPCR_plot_exp1.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now run the script on our second qPCR dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> scripts/qPCR_analysis.R data/qpcr_data2.xlsx qPCR_plot_exp2.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="commenting-and-formatting" class="level3">
<h3 class="anchored">Commenting and formatting</h3>
<p>When writing R scripts, it’s important to use good commenting and formatting practices. This makes your code easier to read and understand, both for yourself and for others who may be reading your code in the future.</p>
<p>Some quick tips:</p>
<ul>
<li>Use <strong>comments</strong> to explain what your code is doing and why. This is especially important for complex code or code that may not be immediately obvious.</li>
<li>Use <strong>consistent indentation and spacing</strong> to make your code more readable. This can help to visually separate different blocks of code and make it easier to follow the flow of the script.</li>
<li>Use <strong>meaningful variable names</strong> that describe the purpose of the variable.</li>
<li><strong>Break up long lines of code</strong> into multiple lines to improve readability. This can help to prevent horizontal scrolling and make it easier to read the code.</li>
<li>Use <strong>functions</strong> to break up your code into smaller, more manageable pieces.</li>
</ul>
<p>Later on we will cover some tools that can help with formatting your code, such as the <code>styler</code> package and the <code>formatR</code> package.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="commenting-and-formatting">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>See if you can add more customisable arguments to the <code>qPCR_analysis.R</code> script. Be sure to comment your edits and use good formatting practices.</p>
<p><strong>SOME IDEAS</strong></p>
<ul>
<li>Allow the user to specify the title of the plot as an argument.</li>
<li>Allow the user to specify the plot file format to be either png or pdf as an argument (hint: use an if statement).</li>
</ul>
</div>
<p><br></p>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Using scripts and programmable functions is a really powerful way to perform your data analysis. Programming languages provide automation, flexibility and reproducibility.</p>
<div class="resources">
<h2 class="anchored">
<i class="fas fa-book"></i> Resources
</h2>
<ul>
<li>The <a href="https://discdown.org/rprogramming/reporting.html">Discdown R Programming book</a> is a comprehensive resource introducing R datatypes, syntax and programming concepts. It is available online for free.</li>
</ul>
</div>
<p><br></p>
<div class="key-points">
<h2 class="anchored">
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>R is a powerful programming language that allows you to automate tasks and write reproducible code.</li>
<li>Control flow functions like if statements and for loops allow you to write code that can adapt to different situations and perform complex tasks.</li>
<li>Functions are reusable blocks of code that perform a specific task.</li>
<li>R scripts allow you to save and share your code, making analyses reproducible and customisable.</li>
</ul>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>